src/array.rs

```diff
@@ -35,48 +36,82 @@ impl PyArray {
         PyArray(obj)
     }
 
-    /// The number of dimensions in the array
+    /// Returns the number of dimensions in the array.
     ///
-    /// https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_NDIM
+    /// Same as [numpy.ndarray.ndim](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.ndim.html)
+    ///
+    /// # Example
+    /// ```
+    /// # extern crate pyo3; extern crate numpy; fn main() {
+    /// use numpy::{PyArray, PyArrayModule};
+    /// let gil = pyo3::Python::acquire_gil();
+    /// let np = PyArrayModule::import(gil.python()).unwrap();
+    /// let arr = PyArray::new::<f64>(gil.python(), &np, &[4, 5, 6]);
+    /// assert_eq!(arr.ndim(), 3);
+    /// # }
+    /// ```
+    // C API: https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_NDIM
     pub fn ndim(&self) -> usize {
         let ptr = self.as_array_ptr();
         unsafe { (*ptr).nd as usize }
     }
 
-    /// dimensions of the array
+    /// Same as [shape](./struct.PyArray.html#method.shape)
     ///
-    /// https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_DIMS
-    pub fn dims(&self) -> Vec<usize> {
-        let n = self.ndim();
-        let ptr = self.as_array_ptr();
-        let dims = unsafe {
-            let p = (*ptr).dimensions;
-            ::std::slice::from_raw_parts(p, n)
-        };
-        dims.into_iter().map(|d| *d as usize).collect()
+    /// Reserved for backward compatibility.
+    #[inline]
+    pub fn dims(&self) -> &[usize] {
+        self.shape()
     }
 
     pub fn len(&self) -> usize {
-        self.dims().iter().fold(1, |a, b| a * b)
+        self.shape().iter().fold(1, |a, b| a * b)
     }
 
-    /// A synonym for PyArray_DIMS, named to be consistent with the ‘shape’ usage within Python.
-    pub fn shape(&self) -> Vec<usize> {
-        self.dims()
+    /// Returns a slice which contains dimmensions of the array.
+    ///
+    /// Same as [numpy.ndarray.shape](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.shape.html)
+    /// # Example
+    /// ```
+    /// # extern crate pyo3; extern crate numpy; fn main() {
+    /// use numpy::{PyArray, PyArrayModule};
+    /// let gil = pyo3::Python::acquire_gil();
+    /// let np = PyArrayModule::import(gil.python()).unwrap();
+    /// let arr = PyArray::new::<f64>(gil.python(), &np, &[4, 5, 6]);
+    /// assert_eq!(arr.shape(), &[4, 5, 6]);
+    /// # }
+    /// ```
+    // C API: https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_DIMS
+    pub fn shape(&self) -> &[usize] {
+        let n = self.ndim();
+        let ptr = self.as_array_ptr();
+        unsafe {
+            let p = (*ptr).dimensions as *mut usize;
```

Comment:
This cast is potentially dengerous.
`dimensions` is a list of `npy_intp` = `Py_intptr_t` = `libc::intptr_t`, which equals `isize` in most platform as you write for `strides`. 

https://github.com/PyO3/pyo3/blob/ff70f45071c39efc2d857f23d78fb6f9ae5656bb/src/ffi3/pyport.rs#L3
https://github.com/rust-lang/libc/blob/72b16d27cce111cbd4b9055d974d0ea70d152347/src/lib.rs#L145

Therefore,

```
fn shape(&self) -> &[isize];
```

is more safer. Since NumPy itself returns its shape with signed integer, `isize` interface can be reasonable while it is not Rust way.
Otherwise, you should check these values are not negative to assure the cast is safe.

Reply:
Hmm...
Yeah I think `isize` is safer too, but I don't think we have to be too nervous about it.
If numpy returns a negative value as matrix dimension, isn't it numpy's bug?

Reply:
>  If numpy returns a negative value as matrix dimension, isn't it numpy's bug?

This makes sense.